sympy/polys/rootoftools.py

```diff
@@ -407,6 +408,33 @@ def _eval_is_imaginary(self):
             return ivl.ax*ivl.bx <= 0  # all others are on one side or the other
         return False  # XXX is this necessary?
 
+    def _sort_roots(roots):
+        """
+        Sorts roots to be in the order that CRootOf uses.
+        Note that for proper sorting, the class methods
+        _real_sorted() and _complexes_sorted() use isolating
+        intervals, and these are typically used in the root
+        finding methods in this class. This method is used
+        internally if that data is not available. It falls
+        back on numerical evaluation if necessary.
+        """
+        real_roots = sorted(r for r in roots if r.is_real)
+
+        try:
+            complex_roots = sorted((r for r in roots if not r.is_real),
+                                key=lambda root: (re(root), im(root)))
+            """
+            Example error
+            TypeError: cannot determine truth value of Relational:
+            re(CRootOf(x**6 - 2*x**4 - 2*x**3 + 1, 3)) <
+                re(CRootOf(x**6 - 2*x**4 - 2*x**3 + 1, 2))
+            """
+        except TypeError:
+            complex_roots = sorted((r for r in roots if not r.is_real),
+                                key=lambda root: (re(root.evalf()), im(root.evalf())))
```

Comment:
I don't think that this fallback should be used. It cannot reliably order the roots. If we get here then it is almost certainly because two or more roots have identical real part but calling `re(root.eval())` will make them seem not equal leading to an arbitrary order.

Is it not possible to continue using the existing code that sorts the roots?

I realise everything here is that the code here is quite convoluted so making it difficult to do these things.

Here is one way to do this:
```python
roots = p.lift().real_roots() # Existing QQ code can sort the roots

subroots = set()
for f, m in p.sqf_list()[1]:
    subroots.update(p.which_roots(roots, real=True))

roots_filtered = [r for r in roots if r not in subroots]
```
We only need to make this square-free for the real_roots case and only because otherwise count_roots does not give us the multiplicities. For the all_roots case it is just:
```
roots_filtered = p.which_roots(p.lift().all_roots(), real=False)
```
Given that functions like `real_roots` and `all_roots` are separate functions rather than being separated by a keyword argument I think it would make sense to have `which_real_roots_sqf` and `which_roots` as separate methods. The `which_real_roots_sqf` function requires the polynomial to be square free. The `which_roots` function does not. A function `which_real_roots()` could do the dance with `sqf_list` that I showed above before calling `which_real_roots_sqf()`.

Reply:
This is great! Thanks for the help.

After some experimenting, I think that even in the all_roots case, we need to get square-free factors. Eg if `f= Poly((x-1) * (x-sqrt(2))**2 * (x-I) * (x+I), x, extension=True)` (which has roots `[1,sqrt(2),sqrt(2), I, -I]`).

Bu it's lift has roots `[-sqrt(2), -sqrt(2), 1, 1, sqrt(2), sqrt(2), -I, -I, I, I]`. The `-sqrt(2)` are filtered out easily. However, in `which_roots`, if we just compare the sum of counts to the degree (as in the all_roots setting), this will loop forever. Thus, I think it makes sense to always return unique roots in `which_roots()`. And then handle multiplicity at the level of the square-free factorization.

In light of that, I think both cases are similar enough that we can keep `which_roots` one function, separated by the `real` argument? The only difference is the root counting. Otherwise, both should return unique elements.

Reply:
> I think that even in the all_roots case, we need to get square-free factors.

Yes, you are right. The multiplicities might not be equal so we can't count the roots with multiplicities except in the square free case where the multiplicity is always 1.